src/redisearch_rs/inverted_index/src/lib.rs

```diff
@@ -367,3 +388,179 @@ pub trait Decoder {
+    /// that this block covers.
+    last_doc_id: t_docId,
+
+    /// The total number of non-unique entries in this block
+    num_entries: usize,
+
+    /// The encoded entries in this block
+    buffer: Vec<u8>,
+}
+
+impl IndexBlock {
+    const SIZE: usize = std::mem::size_of::<Self>();
+
+    /// Make a new index block with the given initial doc ID. This returns the block and how much
+    /// memory grew by.
+    pub fn new(doc_id: t_docId) -> (Self, usize) {
+        let this = Self {
+            first_doc_id: doc_id,
+            last_doc_id: doc_id,
+            num_entries: 0,
+            buffer: Vec::new(),
+        };
+        let buf_cap = this.buffer.capacity();
+
+        (this, Self::SIZE + buf_cap)
+    }
+
+    fn writer(&mut self) -> Cursor<&mut Vec<u8>> {
+        let pos = self.buffer.len();
+        let mut buffer = Cursor::new(&mut self.buffer);
+
+        buffer.set_position(pos as u64);
+
+        buffer
+    }
+}
+
+impl<E: Encoder> InvertedIndex<E> {
+    /// Create a new inverted index with the given encoder. The encoder is used to write new
+    /// entries to the index.
+    pub fn new(encoder: E) -> Self {
+        Self {
+            blocks: Vec::new(),
+            n_unique_docs: 0,
+            encoder,
+        }
+    }
+
+    /// Add a new record to the index and return by how much memory grew. It is expected that
+    /// the document ID of the record is greater than or equal the last document ID in the index.
+    pub fn add_record(&mut self, record: &RSIndexResult) -> std::io::Result<usize> {
+        let doc_id = record.doc_id;
+
+        let same_doc = match (
+            E::ALLOW_DUPLICATES,
+            self.last_doc_id().map(|d| d == doc_id).unwrap_or_default(),
+        ) {
+            (true, true) => true,
+            (false, true) => {
+                // The encoder does not allow writing the same document to the same index twice. This
+                // can happen when the index is created with duplicate tags for example.
+                return Ok(0);
+            }
+            (_, false) => false,
+        };
+
+        // We take ownership of the block since we are going to keep using self. So we can't have a
+        // mutable reference to the block we are working with at the same time.
+        let (mut block, mut mem_growth) = self.take_block(doc_id, same_doc);
+
+        let delta_base = E::delta_base(&block);
+        debug_assert!(
+            doc_id >= delta_base,
+            "documents should be encoded in the order of their IDs"
+        );
+        let delta = doc_id.wrapping_sub(delta_base);
+
+        let delta = match E::Delta::from_u64(delta) {
+            Some(delta) => delta,
+            None => {
+                // The delta is too large for this encoder. We need to create a new block.
+                // Since the new block is empty, we'll start with `delta` equal to 0.
+                let (new_block, block_size) = IndexBlock::new(doc_id);
```

Comment:
I'm not sure that I got the logic here - when will `wrapping_sub(delta_base)` is expected to return None?